### ISSUE TITLE
STM32F1, USB CDC, issues with writing to bulk endpoint.

### ISSUE THREAD

**Issue by krish2487** [libopencm3/libopencm3#1322]
Hello folks,
I am working with a STM32F103RCT6 and trying to implement USB CDC on the same processor. Before I get into the bulk of the problem, a little bit of peripheral information which I think is relevant. The app is structured in layers.

 

- I have a file app.c which has the main function. 
-  main() calls the Init() (present in Init.c) which brings up the board - POST, GPIO setup, peripheral setup and so on. 
-  Init calls LogInit and LogMessage functions (present in Log.c) as part of the bring up sequence. 
-  LogInit calls InitUSB ( present in DriverUSB.c ) and LogMessage calls WriteUSBMessage ( DriverUSB.c ).
-  DriverUSB.c sits on top of LibopenCM3 HAL and calls the necessary functions to initialize the USB peripheral and write messages to the USB peripheral.

Whew, now the problem. The board does not enumerate on the PC. 
Please find the file contents of the aforementioned files.

app.c 
```
#include "app.h"
#include "BSP.h"
#include "Init.h"


#ifndef TEST
	// #define MAIN main
	int main(void)
#else
	// #define MAIN testable_main
	int testable_main(void)
#endif


// int MAIN(int argc, char const *argv[])
{

	Init();

	#ifndef TEST
	while (1)
		{

		}
	#else
	#endif
	/* code */
	return 0;
}

```

Init.c
```
#include "Init.h"
#include "IO.h"
#include "Log.h"

enum Error_codes Init(void)
{

	SysClockInit();
	if (ERROR == IOInit(SYS_STATUS))
		return ERROR;
	IOBlink(POWERUP);
	LogInit(USB);
	LogMessage(USB, INFO_LEVEL, "Initializing System");

	return SUCCESS;
}

```

Log.c

```

#include "Log.h"
#include "DriverUSB.h"

enum Error_codes LogInit(enum LogSubSystem System)
{
	switch (System)
		{
			case USB:
				InitUSB();
				break;
			case FileSystem:
				break;
			case SUBSYSTEMS_END:
				return ERROR;
			}
		return SUCCESS;
	}

	enum Error_codes LogMessage(enum LogSubSystem System, enum LogLevels level, char *message)
{

	WriteUSBMessage(message);

	return SUCCESS;
}

```

DriverUSB.c
```
#include "DriverUSB.h"
#include "stddef.h"
#include "string.h"

#ifndef TEST

	#include <libopencm3/stm32/rcc.h>
	#include <libopencm3/stm32/gpio.h>
	#include <libopencm3/cm3/nvic.h>


#elif TEST
	#include "wrapper.h"

#endif

usbd_device *usbd_dev;
uint8_t usbd_control_buffer[128];

enum usbd_request_return_codes cdcacm_control_request(usbd_device *usbd_dev, struct usb_setup_data *req, uint8_t **buf,
        uint16_t *len, void (**complete)(usbd_device *usbd_dev, struct usb_setup_data *req))
{
	(void)complete;
	(void)buf;
	(void)usbd_dev;

	switch (req->bRequest)
		{
			case USB_CDC_REQ_SET_CONTROL_LINE_STATE:
				{
					/*
					 * This Linux cdc_acm driver requires this to be implemented
					 * even though it's optional in the CDC spec, and we don't
					 * advertise it in the ACM functional descriptor.
					 */
					char local_buf[10];
					struct usb_cdc_notification *notif = (void *)local_buf;

					/* We echo signals back to host as notification. */
					notif->bmRequestType = 0xA1;
					notif->bNotification = USB_CDC_NOTIFY_SERIAL_STATE;
					notif->wValue = 0;
					notif->wIndex = 0;
					notif->wLength = 2;
					local_buf[8] = req->wValue & 3;
					local_buf[9] = 0;
					// usbd_ep_write_packet(0x83, buf, 10);
					return USBD_REQ_HANDLED;
				}
			case USB_CDC_REQ_SET_LINE_CODING:
				if (*len < sizeof(struct usb_cdc_line_coding))
					return USBD_REQ_NOTSUPP;

				return USBD_REQ_HANDLED;
		}
	return USBD_REQ_NOTSUPP;
}

void cdcacm_data_rx_cb(usbd_device *usbd_dev, uint8_t ep)
{
	(void)ep;

	char buf[64];
	int len = usbd_ep_read_packet(usbd_dev, 0x01, buf, 64);

	if (len)
		{
			usbd_ep_write_packet(usbd_dev, 0x82, buf, len);
			buf[len] = 0;
		}
}

void cdcacm_data_tx_cb(usbd_device *usbd_dev, uint8_t ep)
{
	(void)ep;
	(void)usbd_dev;
}

void cdcacm_set_config(usbd_device *usbd_dev, uint16_t wValue)
{
	(void)wValue;

	usbd_ep_setup(usbd_dev, 0x01, USB_ENDPOINT_ATTR_BULK, 64, cdcacm_data_rx_cb);
	usbd_ep_setup(usbd_dev, 0x82, USB_ENDPOINT_ATTR_BULK, 64, NULL);
	usbd_ep_setup(usbd_dev, 0x83, USB_ENDPOINT_ATTR_INTERRUPT, 16, NULL);

	usbd_register_control_callback(
	    usbd_dev,
	    USB_REQ_TYPE_CLASS | USB_REQ_TYPE_INTERFACE,
	    USB_REQ_TYPE_TYPE | USB_REQ_TYPE_RECIPIENT,
	    cdcacm_control_request);
}

enum Error_codes InitUSB(void)
{

	rcc_periph_clock_enable(RCC_USB);
	rcc_periph_reset_pulse(RST_USB);
	usbd_dev =  usbd_init(&st_usbfs_v1_usb_driver, &dev, &config, usb_strings, 3, usbd_control_buffer, sizeof(usbd_control_buffer));
	nvic_set_priority(NVIC_USB_LP_CAN_RX0_IRQ, 2 << 6);
	nvic_enable_irq(NVIC_USB_LP_CAN_RX0_IRQ);
	nvic_enable_irq(NVIC_USB_WAKEUP_IRQ);
	return SUCCESS;
}

enum Error_codes WriteUSBMessage(char *message)
{
	if (message == NULL)
		{
			return ERROR;
		}

	usbd_ep_write_packet(usbd_dev, 0x82, message, strlen(message)  );
	return SUCCESS;
}

// This is the interrupt handler for low priority USB events. Implementing a
// function with this name makes it the function used for the interrupt.
// TODO: Handle the other USB interrupts.
void usb_lp_can_rx0_isr(void)
{

	gpio_toggle(GPIOB, GPIO1);
	usbd_poll(usbd_dev);

}

void usb_wakeup_isr(void)
{
	gpio_toggle(GPIOB, GPIO0);
	usbd_poll(usbd_dev);
}

```
DriverUSB.h

```
#ifndef DRIVERUSB_H
#define DRIVERUSB_H
#include <libopencm3/usb/usbd.h>
#include <libopencm3/usb/cdc.h>
#include "BSP.h"


extern uint8_t usbd_control_buffer[128];
extern usbd_device *usbd_dev;

static const struct usb_device_descriptor dev =
{
	.bLength = USB_DT_DEVICE_SIZE,
	.bDescriptorType = USB_DT_DEVICE,
	.bcdUSB = 0x0200,
	.bDeviceClass = USB_CLASS_CDC,
	.bDeviceSubClass = 0,
	.bDeviceProtocol = 0,
	.bMaxPacketSize0 = 64,
	.idVendor = 0x16D0,
	.idProduct = 0x0F35,
	.bcdDevice = 0x0200,
	.iManufacturer = 1,
	.iProduct = 2,
	.iSerialNumber = 3,
	.bNumConfigurations = 1,
};

/*
 * This notification endpoint isn't implemented. According to CDC spec its
 * optional, but its absence causes a NULL pointer dereference in Linux
 * cdc_acm driver.
 */
static const struct usb_endpoint_descriptor comm_endp[] = {{
		.bLength = USB_DT_ENDPOINT_SIZE,
		.bDescriptorType = USB_DT_ENDPOINT,
		.bEndpointAddress = 0x83,
		.bmAttributes = USB_ENDPOINT_ATTR_INTERRUPT,
		.wMaxPacketSize = 16,
		.bInterval = 255,
	}
};

static const struct usb_endpoint_descriptor data_endp[] = {{
		.bLength = USB_DT_ENDPOINT_SIZE,
		.bDescriptorType = USB_DT_ENDPOINT,
		.bEndpointAddress = 0x01,
		.bmAttributes = USB_ENDPOINT_ATTR_BULK,
		.wMaxPacketSize = 64,
		.bInterval = 1,
	}, {
		.bLength = USB_DT_ENDPOINT_SIZE,
		.bDescriptorType = USB_DT_ENDPOINT,
		.bEndpointAddress = 0x82,
		.bmAttributes = USB_ENDPOINT_ATTR_BULK,
		.wMaxPacketSize = 64,
		.bInterval = 1,
	}
};

static const struct
{
	struct usb_cdc_header_descriptor header;
	struct usb_cdc_call_management_descriptor call_mgmt;
	struct usb_cdc_acm_descriptor acm;
	struct usb_cdc_union_descriptor cdc_union;
} __attribute__((packed)) cdcacm_functional_descriptors =
{
	.header = {
		.bFunctionLength = sizeof(struct usb_cdc_header_descriptor),
		.bDescriptorType = CS_INTERFACE,
		.bDescriptorSubtype = USB_CDC_TYPE_HEADER,
		.bcdCDC = 0x0110,
	},
	.call_mgmt = {
		.bFunctionLength =
		sizeof(struct usb_cdc_call_management_descriptor),
		.bDescriptorType = CS_INTERFACE,
		.bDescriptorSubtype = USB_CDC_TYPE_CALL_MANAGEMENT,
		.bmCapabilities = 0,
		.bDataInterface = 1,
	},
	.acm = {
		.bFunctionLength = sizeof(struct usb_cdc_acm_descriptor),
		.bDescriptorType = CS_INTERFACE,
		.bDescriptorSubtype = USB_CDC_TYPE_ACM,
		.bmCapabilities = 0,
	},
	.cdc_union = {
		.bFunctionLength = sizeof(struct usb_cdc_union_descriptor),
		.bDescriptorType = CS_INTERFACE,
		.bDescriptorSubtype = USB_CDC_TYPE_UNION,
		.bControlInterface = 0,
		.bSubordinateInterface0 = 1,
	}
};

static const struct usb_interface_descriptor comm_iface[] = {{
		.bLength = USB_DT_INTERFACE_SIZE,
		.bDescriptorType = USB_DT_INTERFACE,
		.bInterfaceNumber = 0,
		.bAlternateSetting = 0,
		.bNumEndpoints = 1,
		.bInterfaceClass = USB_CLASS_CDC,
		.bInterfaceSubClass = USB_CDC_SUBCLASS_ACM,
		.bInterfaceProtocol = USB_CDC_PROTOCOL_AT,
		.iInterface = 0,

		.endpoint = comm_endp,

		.extra = &cdcacm_functional_descriptors,
		.extralen = sizeof(cdcacm_functional_descriptors)
	}
};

static const struct usb_interface_descriptor data_iface[] = {{
		.bLength = USB_DT_INTERFACE_SIZE,
		.bDescriptorType = USB_DT_INTERFACE,
		.bInterfaceNumber = 1,
		.bAlternateSetting = 0,
		.bNumEndpoints = 2,
		.bInterfaceClass = USB_CLASS_DATA,
		.bInterfaceSubClass = 0,
		.bInterfaceProtocol = 0,
		.iInterface = 0,

		.endpoint = data_endp,
	}
};

static const struct usb_interface ifaces[] = {{
		.num_altsetting = 1,
		.altsetting = comm_iface,
	}, {
		.num_altsetting = 1,
		.altsetting = data_iface,
	}
};

static const struct usb_config_descriptor config =
{
	.bLength = USB_DT_CONFIGURATION_SIZE,
	.bDescriptorType = USB_DT_CONFIGURATION,
	.wTotalLength = 0,
	.bNumInterfaces = 2,
	.bConfigurationValue = 1,
	.iConfiguration = 0,
	.bmAttributes = 0x80,
	.bMaxPower = 0x32,

	.interface = ifaces,
};

static const char *usb_strings[] =
{
	"ACME Inc",
	"ACME Product #1",
	"AWS TEST1",
};



// extern enum usbd_request_return_codes cdcacm_control_request(usbd_device *usbd_dev, struct usb_setup_data *req, uint8_t **buf, uint16_t *len, void (**complete)(usbd_device *usbd_dev, struct usb_setup_data *req));

// extern void cdcacm_data_rx_cb(usbd_device *usbd_dev, uint8_t ep);

// extern void cdcacm_data_tx_cb(usbd_device *usbd_dev, uint8_t ep);

// extern void cdcacm_set_config(usbd_device *usbd_dev, uint16_t wValue);

enum Error_codes InitUSB(void);

enum Error_codes WriteUSBMessage(char *message);
#endif // DRIVERUSB_H

```

DriverUSB.c/.h files are adapted from the libopenCM3 USB CDC examples for the STM32F1. 

Now the problem I am seeing is here, in the Init.c file.

```
	LogInit(USB);
	LogMessage(USB, INFO_LEVEL, "Initializing System");
```

The board enumerates perfectly fine and I can see a USB CDC device  when I comment out
```
	LogMessage(USB, INFO_LEVEL, "Initializing System");
```

but fails, when I add it. I have tried debugging the code with Ozone and Jlink edu. And it looks like the LogMessage causes  a HardFault Exception.

This is what the call stack looks like 
**reset_handler -> main -> Init -> LogMessage -> WriteUSBMessage -> st_usbfs_ep_write_packet -> st_usbfs_copy_to_pm -> blocking_handler**

st_usbfs_ep_write_packet is present in st_usbfs_core.c
st_usbfs_copy_to_pm is present in st_usbfs_v1.c

AFAICS, something in the definition of my LogMessage is causing the crash in the copy_to_pm function, I suspect it is something about the length of the message.. but strlen should just return the length of the string, I am attempting to write. 

I will be grateful if someone more knowledgeable about the USB Stack implementation can point out what I am doing wrong. Hopefully it is a silly issue from my code.. 

Thanks in advance folks!! and have a wonderful weekend!! :-)

**Comment by manuelbl** [comment]
Can you better describe the behavior you are observing? You write that the device does not enumerate on the PC. Does it mean that the USB device does not appear at all (in the Windows Device Manager) or does it mean that it does appear but not as a COM port?

I've seen that the code toggles a GPIO pin in the USB interrupt handler? How many times does it toggle when you plug it in?

The STM32 F103 series does not have a built-in pull-up resistor on D+. Can you give more details about your hardware and how you have implemented the pull-up resistor?

**Comment by krish2487** [comment]
@manuelbl  
Thank you for the prompt response. The device does not enumerate at all, when the "LogMessage" function is uncommented. 
When I do comment it out, the device shows up as a USB CDC device and I can even connect using a terminal emulator, however it does not print anything. 

When the LogMessage is commented out, the LED toggles, when I connect or disconnect the terminal emulator from the port. 
When the LogMessage is uncommented, the LED does not toggle at all. 

It is a custom board that I have designed and the hardware has been validated quite a while ago. 
<img width="1005" alt="Screen Shot 2021-04-03 at 13 11 40" src="https://user-images.githubusercontent.com/23363406/113476822-34f20b00-947e-11eb-9cc4-58f7de68196e.png">

As you can see R204 is the pull up on the D+ line. IIRC it was a 4K7 pull up.

**Comment by manuelbl** [comment]
I might have spotted the problem. The proper sequence of initialization (related to USB) is:

1. Call `usbd_init()`
2. Register a configuration callback using `usbd_register_set_config_callback()`
3. When the configuration callback is called, create endpoints using `usbd_ep_setup()`
 
Only after step 3 has been completed, `LogMessage` may be used. Before that, it's likely to crash.

As far as I can see, `usbd_register_set_config_callback()` is not called at all. Thus the configuration callback `cdcacm_set_config()` isn't called either and no endpoints are ever set up.

The below changes should fix the problem:

DriverUSB.h
```
extern int is_usb_configured;
```

DriverUSB.c
```
int is_usb_configured = 0;

enum Error_codes InitUSB(void)
{
    ...
    nvic_enable_irq(NVIC_USB_WAKEUP_IRQ);
    usbd_register_set_config_callback(dev, cdcacm_set_config);
    return SUCCESS;
}

void cdcacm_set_config(usbd_device *usbd_dev, uint16_t wValue)
{
    ...
    usbd_register_control_callback(
        usbd_dev,
        USB_REQ_TYPE_CLASS | USB_REQ_TYPE_INTERFACE,
        USB_REQ_TYPE_TYPE | USB_REQ_TYPE_RECIPIENT,
        cdcacm_control_request);
    is_usb_configured = 1;
}
```

Log.c
```
enum Error_codes LogMessage(enum LogSubSystem System, enum LogLevels level, char *message)
{
    if (!is_usb_configured)
        return ERROR;

    WriteUSBMessage(message);
    return SUCCESS;
}
```

**Comment by krish2487** [comment]
@manuelbl 
You are right.. I totally missed calling the usbd_register_set_config_callback function.. 
🤦 
I m reasonably confident that this should fix the issue.. Please do give me some time to check it on the hardware. I am a little away from the hardware atm. 

Nevertheless, thank you for the help and the quick eyes to spot the issue. :-)

**Comment by krish2487** [comment]
Well it does sort of work... and also doesnt.
You were right about the usbd_register_set_config_callback not being called. I did make the changes you suggested and this function is called.

However, the cdcacm_set_config function is never called. I.e the endpoints are never setup. and there is nothign printed on the serial port since it returns an error. I did step and verify that the cdcacm_set_config is never called. it just returns from the register_set_config_callback function. 

There is one change I did to the above mentioned changes, you have suggested.
You have the cdcacm_set_config defined after the InitUSB function. I moved it before the InitUSB function, i.e as in the original source file I posted above.

**Comment by krish2487** [comment]
Ok.. a little more debugging..
I modified the InitUSB function to so

```
enum Error_codes InitUSB(void)
{

	rcc_periph_clock_enable(RCC_GPIOA);
	rcc_periph_clock_enable(RCC_USB);
	rcc_periph_reset_pulse(RST_USB);
	usbd_dev =  usbd_init(&st_usbfs_v1_usb_driver, &dev, &config, usb_strings, 3, usbd_control_buffer, sizeof(usbd_control_buffer));
	usbd_register_set_config_callback(usbd_dev, cdcacm_set_config);
	nvic_set_priority(NVIC_USB_LP_CAN_RX0_IRQ, 2 << 6);
	nvic_enable_irq(NVIC_USB_LP_CAN_RX0_IRQ);
	nvic_enable_irq(NVIC_USB_WAKEUP_IRQ);
	if (is_usb_configured == 0)
		return ERROR;
	return SUCCESS;
}
```
to check if the cdcacm_set_config callback is actually being called and the Endpoints are being set.
The function returns ERROR. meaning that the 
```
is_usb_configured
```
variable is never being set. This makes me suspect that the cdcacm_set_config is not being called at all, since that is the only place we are actually setting the variable. I stepped thru the code with the debugger and can confirm this.

@manuelbl  Any suggestions on what I can try next?

**Comment by manuelbl** [comment]
You seem to expect that `cdcacm_set_config()` is immediately called from `usbd_register_set_config_callback()`. That's not the case.  `cdcacm_set_config()` will be called once Windows decides to configure the USB device.

Therefore, you will not be able to send any log messages right after startup. You should probably remove the statement `LogMessage(USB, INFO_LEVEL, "Initializing System");` as it cannot possibly work. Log output that is generated later (probably about 2s after you have plugged in the device) will be successfully sent.

So what's the overall behavior now? Does the device appear as a COM port in Windows but not generate any output? Or does it no longer appear as a COM port? I hope that at least the hard fault has been fixed.

**Comment by krish2487** [comment]
Oh my mistake, I assumed that it would be called immediately. I didnt realise it would depend on the host system to actually instantiate the endpoints.

That is correct. The device shows up as a COM port but does not send any messages. 
Would introducing a delay of about 5 seconds give it enough time to enumerate?? 
between the set_config_callback and the cdcacm_set_config being called?? Less elegant but I guess I can refactor it into the code into something more practical if such a solution works...

**Comment by krish2487** [comment]
I can confirm that introducing a delay of about 5 seconds between the InitUSB and LogMessage indeed does allow the host to configure the device correctly and the initial message is printed to the console as well.

@manuelbl  I owe a thanks and beer to you for being so astute and being able to pinpoint the source of the problem at hand - me . :-)

Thank you again fro your help and suggestions!! I am going to close this issue now and continue on my travels.. :-P
